src/SimulationMap.cu

```diff
@@ -1,13 +1,241 @@
+                                                        left_coordinates, current_face.id, polyhedron);
+            int left_mapnode_id = find_nearest_mapnode_to_point(left_coordinates);
+            current_node.set_left(&nodes[left_mapnode_id]);
+        }
+        if(current_node.get_bottom() == nullptr)
+        {
+            SpacePoint bottom_coordinates = current_node.get_coordinates() - top_direction;
+            bottom_coordinates = get_projected_vector_end(nodes[current_node_id].get_coordinates(),
+                                                          bottom_coordinates, current_face.id, polyhedron);
+            int bottom_mapnode_id = find_nearest_mapnode_to_point(bottom_coordinates);
+            current_node.set_bottom(&nodes[bottom_mapnode_id]);
+        }
+        if(current_node.get_right() == nullptr)
+        {
+            SpacePoint right_coordinates = relative_point_rotation(current_node.get_coordinates(),
+                                                                   current_node.get_coordinates() + top_direction,
+                                                                   current_face.normal, M_PI / 2);
+            right_coordinates = get_projected_vector_end(nodes[current_node_id].get_coordinates(),
+                                                         right_coordinates, current_face.id, polyhedron);
+            int right_mapnode_id = find_nearest_mapnode_to_point(right_coordinates);
+            current_node.set_right(&nodes[right_mapnode_id]);
+        }
+    }
+}
 
 __device__ SimulationMap::~SimulationMap()
 {
-    delete[] nodes;
+    free(nodes);
+}
+
+
+__device__ int SimulationMap::find_nearest_mapnode_to_point(SpacePoint point_coordinates) const
+{
+    int nearest_mapnode_id = 0;
+    for(int neighbor = 0; neighbor < n_of_nodes; ++neighbor)
+    {
+        if (get_distance(nodes[neighbor].get_coordinates(), point_coordinates) <
+                get_distance(nodes[nearest_mapnode_id].get_coordinates(), point_coordinates))
+        {
+            nearest_mapnode_id = neighbor;
+        }
+    }
+    return nearest_mapnode_id;
+}
+
+__device__ int SimulationMap::get_neighbor_mapnode_id(int current_node_id, SpacePoint neighbor_coordinates,
+                                                      SpacePoint **nodes_directions, SpacePoint top_direction)
+{
+    int current_face_id = nodes[current_node_id].get_face_id();
+    neighbor_coordinates = get_projected_vector_end(nodes[current_node_id].get_coordinates(), neighbor_coordinates,
+                                                    current_face_id, polyhedron);
+    int next_face_id = polyhedron->find_face_id_by_point(neighbor_coordinates);
+    int neighbor_mapnode_id = find_nearest_mapnode_to_point(neighbor_coordinates);
+    if (current_face_id == nodes[neighbor_mapnode_id].get_face_id() &&
+        get_distance(nodes[neighbor_mapnode_id].get_coordinates(), neighbor_coordinates) < mapnode_dist / 100)
```

Comment:
I'm kind of worried about the `100` constant here. What does it mean? Should it be here?

Reply:
I suppose using `eps` constant is more correct in this unequality